### PR TITLE
add some new parser hooks to the parser to support FIR types

### PR DESCRIPTION
**Pull request by schweitzpgi**
allow access to '?', '*', and optional '{' and '}' tokens.

### CODE REVIEW

File: include/mlir/IR/DialectImplementation.h [hunk 1]
```diff
@@ -188,6 +188,12 @@ class DialectAsmParser {
   /// Parse a `}` token.
   virtual ParseResult parseRBrace() = 0;
 
+  /// Parse a '{' token if present
+  virtual ParseResult parseOptionalLBrace() = 0;
+
```
Comment (River707) [inline include/mlir/IR/DialectImplementation.h]: Can you keep the same ordering as the .cpp file?

File: include/mlir/IR/DialectImplementation.h [hunk 2]
```diff
@@ -264,6 +270,12 @@ class DialectAsmParser {
   /// Parse a `...` token if present;
   virtual ParseResult parseOptionalEllipsis() = 0;
 
+  /// Parse a `*` token.
+  virtual ParseResult parseOptionalStar() = 0;
+
```
Comment (River707) [inline include/mlir/IR/DialectImplementation.h]: Can you try to keep these sorted alphabetically?

File: lib/Parser/Parser.cpp [hunk 3]
```diff
@@ -596,6 +606,16 @@ class CustomDialectAsmParser : public DialectAsmParser {
     return success(parser.consumeIf(Token::r_square));
   }
 
+  /// Parses a '*' if present.
+  ParseResult parseOptionalStar() override {
```
Comment (River707) [inline lib/Parser/Parser.cpp]: There is a somewhat loose attempt to keep these sorted alphabetically.

File: include/mlir/IR/DialectImplementation.h (hunk 1 above)
Comment (schweitzpgi) [reply include/mlir/IR/DialectImplementation.h]: You bet.

File: lib/Parser/Parser.cpp (hunk 3 above)
Comment (schweitzpgi) [reply lib/Parser/Parser.cpp]: Can you clarify this a bit more? I'm seeing a list of open-thingy, close-thingy pairs with non-optional and optional variants intermingled. But Less, Greater, LParen, RParen, ... Ellipsis doesn't seem to be alphabetical to me.

Maybe just put Question before Star?

Thanks.

Comment (River707) [reply lib/Parser/Parser.cpp]: Yes. That's why I said "loose". Ellipsis is definitely in the wrong spot.

In general I keep the parseL and parseR next to each other, same with the optional. I'll clean up the outliers in a followup, thanks for pointing them out!

**Comment by schweitzpgi** [comment]
OK, done.

**Pull request closed without merging**